consensus/hotstuff/voteaggregator/vote_aggregator.go

```diff
@@ -2,390 +2,170 @@ package voteaggregator
-
-	// try to build the QC with existing votes
-	newQC, built, err := va.tryBuildQC(block.BlockID)
-	if err != nil {
-		return nil, false, fmt.Errorf("could not build QC: %w", err)
-	}
-
-	return newQC, built, nil
-}
-
-func (va *VoteAggregator) isBlockReceived(block *model.Block) bool {
-	// We won't create voting status until the block is received. Therefore, if voting status
-	// has been created for a certain block, the block must have been received.
-	_, exist := va.blockIDToVotingStatus[block.BlockID]
-	return exist
 }
 
-// StoreProposerVote stores the vote for a block that was proposed.
-func (va *VoteAggregator) StoreProposerVote(vote *model.Vote) bool {
-	// check if the proposer vote is for a view that has already been pruned (and is thus stale)
-	if va.isVoteStale(vote) { // cannot store vote for already pruned view
-		return false
-	}
-	// add proposer vote, return false if it exists
-	_, exists := va.proposerVotes[vote.BlockID]
-	if exists {
-		return false
-	}
-	va.proposerVotes[vote.BlockID] = vote
-	return true
-}
-
-// BuildQCOnReceivedBlock will attempt to build a QC for the given block when there are votes
-// with enough stakes.
-// It's idempotent. Meaning, calling it again with the same block returns the same result.
-// It assumes that the proposer's vote has been stored by calling StoreProposerVote
-// It assumes the block has been validated.
-// It returns (qc, true, nil) if a QC is built
-// It returns (nil, false, nil) if not enough votes to build a QC.
-// It returns (nil, false, nil) if the block is stale
-// It returns (nil, false, err) if there is an unknown error
-func (va *VoteAggregator) BuildQCOnReceivedBlock(block *model.Block) (*flow.QuorumCertificate, bool, error) {
-	// return the QC that was built before if exists
-	oldQC, exists := va.createdQC[block.BlockID]
-	if exists {
-		return oldQC, true, nil
-	}
-
-	// if the block is stale, we just return nil
-	if va.isBlockStale(block) {
-		return nil, false, nil
-	}
-
-	// proposer vote is the first to be accumulated
-	proposerVote, exists := va.proposerVotes[block.BlockID]
-	if !exists {
-		// proposer must has been stored before, otherwise it's a bug.
-		// cannot build qc if proposer vote does not exist
-		return nil, false, fmt.Errorf("could not get proposer vote for block: %x, blockView: %v, highestPrunedView: %v", block.BlockID, block.View, va.highestPrunedView)
-	}
-
-	// accumulate proposer's vote first to ensure proposer's vote is always included in the QC.
-	// a consensus node is rewarded by having proposed block being included in the finalized chain.
-	// Including the proposer's vote first guarantees that the proposer will get rewarded if the block
-	// is finalized. It incentivizes an honest leader to propose block.
-	valid, err := va.validateAndStoreIncorporatedVote(proposerVote, block)
+func (va *VoteAggregator) processPendingVote(vote *model.Vote) error {
+	lazyInitCollector, err := va.collectors.GetOrCreateCollector(vote.View, vote.BlockID)
```

Comment:
We delay initialization since we don't know if we got first vote or the block itself, so we will just create collector by demand.

> If we want to know whether the collector was created by this call or other call, we could let GetOrCreateCollector return (VoteCollector, bool, error) instead. The additional bool indicates if was created by this call:

That's another way of doing, I just prefer a separate struct comparing to multiple args without obvious context.